### PR TITLE
refactor(cli): Lazy load config

### PR DESCRIPTION
**Pull request by epage**
This is trying to clarify `-C` support when it is implemented in #10952 

Cargo currently has two initialization states for Config, `Config::default` (process start) and `config.configure` (after parsing args).  The most help we provide for a developer touching this code is a giant `CAUTION` comment in one of the relevant functions.

Currently, #10952 adds another configuration state in the middle where the `current_dir` has been set.

The goal of this PR is to remove that third configuration state by
- Lazy loading `Config::default` so it can be called after parsing `-C`
- Allowing `-C` support to assert that the config isn't loaded yet to catch bugs with it

The hope is this will make the intent of the code clearer and reduce the chance for bugs.

In doing this, there are two intermediate refactorings
- Make help behave like other subcommands
  - Before, we had hacks to intercept raw arguments and to intercept clap errors and assume what their intention was to be able to implement our help system.
  - This flips it around and makes help like any other subcommand,
simplifying cargo initialization.
  - We had to upgrade clap because this exposed a bug where `Command::print_help` wasn't respecting `disable_colored_help(true)`
- Delay fix's access to config

Personally, I also find both changes make the intent of the code clearer.

To review this, I recommend looking at the individual commits.   As this is just refactors, this has no impact on testing.

### CODE REVIEW

File: src/bin/cargo/commands/mod.rs [hunk 1]
```diff
@@ -52,6 +53,7 @@ pub fn builtin_exec(cmd: &str) -> Option<fn(&mut Config, &ArgMatches) -> CliResu
         "doc" => doc::exec,
         "fetch" => fetch::exec,
         "fix" => fix::exec,
+        "help" => help::exec,
         "generate-lockfile" => generate_lockfile::exec,
         "git-checkout" => git_checkout::exec,
```
Comment (weihanglo) [inline src/bin/cargo/commands/mod.rs:58]: Looks like it is in an alphabetical order.

```suggestion
        "generate-lockfile" => generate_lockfile::exec,
        "git-checkout" => git_checkout::exec,
        "help" => help::exec,
```

File: src/cargo/ops/fix.rs [hunk 2]
```diff
@@ -333,20 +333,22 @@ to prevent this issue from happening.
...
 
-/// Entry point for `cargo` running as a proxy for `rustc`.
-///
-/// This is called every time `cargo` is run to check if it is in proxy mode.
+/// Provide the lock address when running in proxy mode
 ///
-/// Returns `false` if `fix` is not being run (not in proxy mode). Returns
-/// `true` if in `fix` proxy mode, and the fix was complete without any
+/// Returns `None` if `fix` is not being run (not in proxy mode). Returns
+/// `Some(...)` if in `fix` proxy mode, and the fix was complete without any
```
Comment (weihanglo) [inline src/cargo/ops/fix.rs:339]: Part of this comment is not true anymore. Perhaps move them to `fix_exec_rustc`?

File: src/bin/cargo/cli.rs [hunk 3]
```diff
@@ -463,6 +465,34 @@ See 'cargo help <command>' for more information on a specific command.\n",
         .subcommands(commands::builtin())
 }
 
+pub struct LazyConfig {
```
Comment (weihanglo) [inline src/bin/cargo/cli.rs:468]: Could we add some rustdoc for this type? (something like intention and usage)

**Review by weihanglo — COMMENTED** [review]
Clever solution for lazy loading! Thank you!

**Review by weihanglo — APPROVED** [review]
👍🏾

**Comment by lqd** [comment]
I think this PR introduced this warning https://github.com/rust-lang/cargo/runs/8128683225?check_suite_focus=true#step:10:257 ?

**Comment by weihanglo** [comment]
> I think this PR introduced this warning https://github.com/rust-lang/cargo/runs/8128683225?check_suite_focus=true#step:10:257 ?

It might eventually be used in this patch <https://github.com/rust-lang/cargo/pull/10952#discussion_r946915045>, though I cannot guarantee when the author will pick it up. If the warning is annoying, we can remove the function for sure.

**Comment by lqd** [comment]
Oh not annoying, I was mostly surprised as I was seeing "deny warnings" features and wondered if CI had an issue checking them 😅

**Pull request merged**